pkg/health/checker/checker.go

```diff
@@ -83,38 +83,87 @@ func (c consulHealthChecker) WatchNodeService(
 	}
 }
 
+func publishLatestHealth(inCh <-chan api.KVPairs, quitCh <-chan struct{}, resultCh chan []*health.Result) chan error {
+	errCh := make(chan error)
+
+	go func() {
+		var listed api.KVPairs
+		var ok bool
+		var err error
+
+		for {
+			// We don't have a value, fetch a fresh one or skip
+			select {
+			case listed, ok = <-inCh:
+				if !ok {
+					// channel closed
+					return
+				}
+			case <-quitCh:
+				return
+			}
+
+			results := make([]*health.Result, 0, len(listed)) // allocate a new return slice for each watch
+			results, err = kvpsToResult(listed)
+			if err != nil {
+				select {
+				case errCh <- err:
+					// The most recent update is in error.
+					// We go back to the start in this case
+					continue
+				case <-quitCh:
+					return
+				}
+			}
+
+			// here, we prepare to write the value.
+			// First we drain the resultChan of any stale health results
+			select {
+			case _, ok = <-resultCh:
+				if !ok {
+					return
+				}
+			default:
+			}
+			// Now we check the quit chan and try to write to our resultCh
+			select {
+			case <-quitCh:
+				return
+			default:
+			}
+			resultCh <- results
```

Comment:
this should be inside the `select` above, otherwise there's a race where quit might fire before you've written resultCh. Also, you `default` is unnecessary if the channel is buffered as we expect but i suppose it's harmless